client/cmd/resource/plan.go

```diff
@@ -0,0 +1,233 @@
+		PreRunE: planCmd.PreRunE,
+		RunE:    planCmd.RunE,
+	}
+
+	planCmd.inject(cmd)
+	return cmd
+}
+
+func (p *planCommand) inject(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&p.configFilePath, "config", "c", config.EmptyPath, "File path for client configuration")
+
+	cmd.Flags().StringVarP(&p.gitProvider, "git-provider", "p", os.Getenv("GIT_PROVIDER"), "selected git provider used in the repository")
+	cmd.Flags().StringVarP(&p.gitURL, "git-host", "h", os.Getenv("GIT_HOST"), "Git host based on git provider used in the repository")
+	cmd.Flags().StringVarP(&p.gitToken, "git-token", "t", os.Getenv("GIT_TOKEN"), "Git token based on git provider used in the repository")
+
+	cmd.Flags().StringVarP(&p.projectID, "project-id", "I", os.Getenv("GIT_PROJECT_ID"), "Determine which project will be checked")
+
+	cmd.Flags().StringVarP(&p.output, "output", "o", "./resource.json", "File path for output of plan")
+	cmd.Flags().BoolVarP(&p.verbose, "verbose", "v", false, "Print details related to operation")
+}
+
+func (p *planCommand) PreRunE(_ *cobra.Command, args []string) error {
+	var err error
+
+	p.destinationRef = args[0]
+	p.sourceRef = args[1]
+	p.logger.Info("[plan] compare: `%s` ← `%s`", p.destinationRef, p.sourceRef)
+	p.clientConfig, err = config.LoadClientConfig(p.configFilePath)
+	if err != nil {
+		return err
+	}
+
+	switch p.gitProvider {
+	case "gitlab":
+		p.repository, err = gitlab.NewAPI(p.gitURL, p.gitToken)
+	case "github":
+		p.repository, err = github.NewAPI(p.gitURL, p.gitToken)
+	default:
+		return errors.New("unsupported git provider, we currently only support: [github,gitlab]")
+	}
+	if err != nil {
+		return err
+	}
+
+	p.specReadWriter, err = specio.NewResourceSpecReadWriter(afero.NewOsFs())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *planCommand) RunE(_ *cobra.Command, _ []string) error {
+	ctx := context.Background()
+	diffs, err := p.repository.CompareDiff(ctx, p.projectID, p.sourceRef, p.destinationRef)
+	if err != nil {
+		return err
+	}
+
+	directories := providermodel.Diffs(diffs).GetAllDirectories(p.appendDirectory)
+	p.logger.Info("resource plan found changed in directories: %+v", directories)
+
+	var plans plan.Plans
+	for _, directory := range directories {
+		var resourcePlan *plan.Plan
+		resourcePlan, err = p.describePlanFromDirectory(ctx, directory)
+		if err != nil {
+			return err
+		}
+		plans = append(plans, resourcePlan)
+	}
+
+	mergedPlans := plans.MergeMigrateOperation()
+	if p.verbose {
+		for i := range mergedPlans {
+			msg := fmt.Sprintf("[%s] plan operation %s for %s %s", mergedPlans[i].NamespaceName, mergedPlans[i].Operation, mergedPlans[i].Kind, mergedPlans[i].KindName)
+			if mergedPlans[i].OldNamespaceName != nil {
+				msg += " with old namespace: " + *mergedPlans[i].OldNamespaceName
+			}
+			p.logger.Info(msg)
+		}
+	}
+	return p.saveFile(mergedPlans)
+}
+
+func (p *planCommand) describePlanFromDirectory(ctx context.Context, directory string) (*plan.Plan, error) {
```

Comment:
see if we break it down to smaller functions, the for loops & if else conditions normally we can look into keeping into smaller private functions & we can avoid if else chaining